torch/_inductor/triton_bundler.py

```diff
@@ -0,0 +1,234 @@
+import dataclasses
+import logging
+import os
+import uuid
+from pathlib import Path
+from typing import List, Optional, Tuple
+
+from torch._dynamo.utils import counters, dynamo_timed
+from torch._utils_internal import justknobs_check
+
+from .runtime.runtime_utils import triton_cache_dir
+
+
+log = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass(frozen=True)
+class TritonBundleEntry:
+    """
+    When we have compiled a triton kernel, we take note of that kernel by
+    its triton generated hash, its device, and where this kernel is located.
+    This is the minimum information we can use to later retrieve this kernel
+    from file system.
+    """
+
+    kernel_hash: str
+    device: int
+    directory: str
+
+
+@dataclasses.dataclass(frozen=True)
+class TritonKernelArtifact:
+    """
+    Artifact for an individual kernel converted to bytes.
+    Bytes could be a cubin, json, ttir, or ttgir.
+    """
+
+    filename: str
+    payload: bytes = dataclasses.field(repr=False)  # Do not display binary
+
+
+@dataclasses.dataclass(frozen=True)
+class TritonKernelArtifacts:
+    """
+    Collection of artifacts for a particular kernel.
+    """
+
+    kernel_hash: str
+    device: int
+    artifacts: List[TritonKernelArtifact]
+
+
+@dataclasses.dataclass(frozen=True)
+class TritonBundlerMetadata:
+    """
+    Metadata used for instrumentation
+    """
+
+    cached_kernel_names: List[str]
+
+
+class TritonBundler:
+    """
+    Lightweight Triton Kernel bundler that notes each time we compile a triton
+    kernel. When collect is called, converts all the previously noted kernels and
+    their artifacts into a structured bytes blob, and later when write is called
+    it writes this structured blob back to file system.
+
+    Intended Life cycle:
+    - TritonBundler.begin_compile is called when we start compiling in Inductor
+    - TritonBundler.put is called each time a Triton Kernel is compiled
+    - TritonBundler.collect is called when a cache entry is being generated
+    - TritonBundler.read_and_emit is called when a cache entry is read
+    """
+
+    _entries: Optional[List[TritonBundleEntry]] = None
+
+    # __grp__kernel_name.json contains metadata with source code paths
+    # we use this as sentinal value for search and replace
+    _REPLACE_BYTES: bytes = b"[REPLACE]"
```

Comment:
There will be some terrifying SEV when this mucks over some unrelated byte string that just *happens* to be `[REPLACE]`. One way to protect against this is to assert that there are no unrelated occurrences of this string before we munge file paths.